src/Helpers/DataResolver.php

```diff
@@ -0,0 +1,191 @@
+<?php
+
+namespace Firesphere\SolrSearch\Helpers;
+
+use Exception;
+use SilverStripe\Core\ClassInfo;
+use SilverStripe\ORM\ArrayList;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\FieldType\DBField;
+use SilverStripe\ORM\SS_List;
+use SilverStripe\View\ArrayData;
+
+class DataResolver
+{
+    /**
+     * @var DataObject|ArrayList|SS_List|DBField
+     */
+    protected $component;
+    /**
+     * @var array
+     */
+    protected $columns = [];
+    /**
+     * @var mixed|string|null
+     */
+    protected $columnName = '';
+    /**
+     * Supported object types
+     * @var array map of objects to methods
+     */
+    private static $objTypes = [
+        DataObject::class => 'DataObject',
+        ArrayData::class  => 'ArrayData',
+        SS_List::class    => 'List',
+        DBField::class    => 'Field'
+    ];
+
+    /**
+     * DataResolver constructor.
+     * @param DataObject|ArrayList|SS_List $component
+     * @param array|string $columns
+     */
+    public function __construct($component, $columns = [])
+    {
+        if (!is_array($columns)) {
+            $columns = array_filter(explode('.', $columns));
+        }
+        $this->columns = $columns;
+        $this->component = $component;
+        $this->columnName = $this->columns ? array_shift($this->columns) : null;
+    }
+
+    /**
+     * @param DataObject|ArrayData|SS_List|DBField $obj
+     * @param array|string $columns
+     *
+     * @return mixed
+     * @throws Exception
+     */
+    public static function identify($obj, $columns = [])
+    {
+        foreach (self::$objTypes as $type => $method) {
+            if ($obj instanceof $type) {
+                $method = 'resolve' . $method;
+
+                return (new self($obj, $columns))->{$method}();
+            }
+        }
+
+        throw new Exception(sprintf('Class: %s, is not supported.', ClassInfo::shortName($obj)));
+    }
+
+    /**
+     * @param DataObject|ArrayData|SS_List $component
+     * @param array $columns
+     *
+     * @return void
+     * @throws Exception
+     */
+    protected function cannotIdentifyException($component, $columns = [])
+    {
+        throw new Exception(
+            sprintf(
+                'Cannot identify, "%s" from class "%s"',
+                implode('.', $columns),
+                ClassInfo::shortName($component)
+            )
+        );
+    }
+
+    /**
+     * Resolves a Single field in the database.
+     * @return mixed
+     * @throws Exception
+     */
+    protected function resolveField()
+    {
+        if ($this->columnName) {
+            if ($this->component->hasMethod($this->columnName)) {
+                $method = $this->columnName;
+            } elseif ($this->component->hasMethod("get{$this->columnName}")) {
+                $method = "get{$this->columnName}";
+            } else {
+                throw new Exception(
```

Comment:
Ignore CodeClimate mostly, unless you think you can resolve it. Often it's being too strict :)